add conditional functionalization

bdhirsh opened the pull request:
New PR. I tested this manually by running:
```
(/scratch/hirsheybar/pytorch-functionalization-env) hirsheybar@devfair017:/scratch/hirsheybar/pytorch-functionalization/xla$ cat tmp6.py
import torch
from functorch import vmap

def f(x):
    return torch.diag_embed(x)


a = torch.ones(2, 2)
g = vmap(f)
out = g(a)
(/scratch/hirsheybar/pytorch-functionalization-env) hirsheybar@devfair017:/scratch/hirsheybar/pytorch-functionalization/xla$ python tmp6.py
tmp6.py:5: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::copy. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at  /raid/hirsheybar/pytorch-functionalization/functorch/functorch/csrc/BatchedFallback.cpp:106.)
  return torch.diag_embed(x)
tmp6.py:5: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::diagonal_scatter. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at  /raid/hirsheybar/pytorch-functionalization/functorch/functorch/csrc/BatchedFallback.cpp:106.)
```

Before, the vmap call would error due to the inplace `copy_()` call inside of `diag_embed`. Now, the functionalization pass is:
-  calling into `diagonal_scatter`, which is used to go from the "view" back to its "base" tensor
- calling `copy` instead of `copy_`

I originally tried adding a logging tensor test for this in https://github.com/pytorch/pytorch/pull/66101, but unfortunately the `diagonal` call can't be seen by the LoggingTensor (this could probably fixed if/once some version of modal LoggingTensor becomes usable in core). I did add a test checking that the `copy_()` gets converted to a `copy()` call though.

bdhirsh commented:
I'm not actually sure if the fixes in https://github.com/pytorch/pytorch/pull/68269 are required for this to land, but they might be - I'll wait for CI to tell me. I independently fixed an issue with the way I was setting up TLS in this PR that broke some combinations of tests.

zou3519 commented:
Another good test case for functionalization is block_diag. I'm not sure we have any other "forward pass" test cases for DECOMPOSE_FUNCTIONAL, most of the problems we have are with backward formulas that do in-place

bdhirsh commented:
All of the issues with `diag_embed` are fixed, but I encountered a bug around `TensorList` usage when functionalizing `block_diag`. I have a patch here: https://github.com/pytorch/pytorch/pull/68395.

Also, I see that the `test_vmapjvp` tests still fail for both ops - my understanding of vmap-vjap is still loose, but if the idea is that we're effectively calling running `grad()` first and then `vmap()`, then it makes sense that conditional functionalization won't run, since we decompose from the `grad()` call first. I'm leaving those tests as expected failures for now, but maybe we should consider:

(1) also registering `DECOMPOSE_FUNCTIONAL` kernels to the functorch grad key (this has the downside of unnecessarily making some `grad(f)` calls slower)

(2) Trying to do something fancier/hackier if we're not okay with the perf change from (1)

bdhirsh commented:
Followup on dealing with `vmap(grad(torch.diag_embed))`. I think we agree about these two cases:
```
f1 = grad(torch.diag_embed) // this should NOT functionalize diag_embed

f2 = grad(vmap(torch.diag_embed)) // this SHOULD functionalize diag_embed
```

In both cases, we technically have all of the information that we need in order to make that decision, at the time that we dispatch to the `FuncTorchGradWrapper` kernel. That's because even though each transform is run eagerly, we know what the list of transforms is up front. So if we check the layer stack and see vmap further down, we know that we should run functional decomposition instead of regular decomposition.

It would look something like this:
```
void decompose_functional_if_vmap_is_active(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
    for (auto layer : dynamicLayerStackAccessor()) {
        if (layer.key() == c10::DispatchKey::FuncTorchBatched) {
            // if vmap is on the stack of transforms, run functionalization
            decompose_functional(op, stack);
            return;
        }
    }
    // otherwise, go through the standard autograd decomposition
    c10::Dispatcher::singleton().redispatchBoxed(...);
}
```

bdhirsh commented:
Looks like CI is passing now that https://github.com/pytorch/pytorch/pull/68395 has landed.

zou3519 commented on functorch/csrc/BatchRulesDecompositions.cpp:
```diff
@@ -7,12 +7,91 @@
+at::Tensor sync_and_unwrap_functional_output(at::Tensor out_functional) {
+  TORCH_INTERNAL_ASSERT(at::functionalization::impl::isFunctionalTensor(out_functional));
+  auto out_wrapper_impl = at::functionalization::impl::unsafeGetFunctionalWrapper(out_functional);
+  // sigh.. we can't just call sync(). sync() only runs if the current tensor is "not up to date".
+  // Where "not up to date" is defined as "has any aliases".
+  // Unfortunately at this point, any aliases to `out_functional` will have been deallocated,
+  // So we need to effectively force the sync.
```
Why is "not up to date" defined as "has any aliases"?

bdhirsh replied:
I can tell you why `sync()` is set up like that, but you also just made me realize another issue that will pop up in `functionalize()`, and likely in the LazyTensor/XLA integration.

Every non-view/non-inplace operation during the functionalization pass needs to check if its tensor inputs need to be synced before performing the operation, so they don't perform the op on stale data. The requirements for syncing are:
- Does the current tensor have an aliases
- If so, do any of those aliases have a more recent version of the data (checked by checking the `generation_` of the tensor).

There's a constraint though: creating an alias of an existing tensor shouldn't mutate any data, otherwise we'd have a race condition. Ed originally brought this up [here](https://github.com/pytorch/pytorch/pull/64432#discussion_r704488724), and the solution we came to was that we could check if a `FunctionalTensorWrapper` has any aliases by checking its storage's use_count.

So why does that cause problems if we try to call `sync()` here? The logic above doesn't run until after the program has finished, which means any aliases generated inside of the program/function will have been deallocated and the use_count will be 1 again.

Maybe a `force_sync()` or `sync(force=true)` would be a little more readable. Although it only really needs to get used here, so I didn't see a point in creating a one-off API for it.

It just occurred to me that this will cause issues for other uses of functionalization though. If you consider a program like:
```
import torch
from functorch import make_fx, grad, vmap, functionalize

def bar(y):
    y_view = y.view(-1)
    y_view.add_(1)

def foo(x):
    bar(x)
    z = x + 1  # the pass SHOULD sync the updates to x here, but it doesn't, because it sees use_count == 1.
    return z

# prints tensor([3., 3.])
print(foo(torch.ones(2)))

# ... but this prints tensor([2., 2.])
print(functionalize(foo)(torch.ones(2)))
```

I don't think we'll need to fix this problem immediately, but it'll come up soon. I'll need to think about it more... but in the worst case, we might have to allow and clearly document that race condition above.

bdhirsh replied:
actually, maybe we can just get away with not checking for aliasing at all, and only check the generation count.

zou3519 replied:
Checking my understanding: after https://github.com/pytorch/pytorch/pull/68881 can we replace the following
```
  out_wrapper_impl->apply_updates();
  out_wrapper_impl->regenerate_from_base();
```
with `out_wrapper_impl->sync_()`? Since we've updated the `is_up_to_date()` logic?

bdhirsh replied:
yep! That's actually why I'm waiting to land pytorch/pytorch#68881 first - then we can make the change you suggested.

zou3519 commented on functorch/csrc/BatchRulesDecompositions.cpp:
```diff
@@ -7,12 +7,91 @@
+  return out_unwrapped;
+}
+
+at::TensorList sync_and_unwrap_functional_output(const c10::List<at::Tensor>& t_list) {
+  std::vector<at::Tensor> outputs(t_list.size());
+  for (const auto i : c10::irange(t_list.size())) {
+    outputs[i] = sync_and_unwrap_functional_output(t_list[i]);
+  }
+  return outputs;
+}
```
This shouldn't return a TensorList. TensorList is not owning so it'll end up pointing to nothing once `outputs` goes out of scope

bdhirsh replied:
Thanks, yep. This is the same form of bug that I had to fix in core last week.

zou3519 commented on functorch/csrc/BatchRulesDecompositions.cpp:
```diff
@@ -131,6 +210,8 @@ TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
   OP_DECOMPOSE2(unflatten, int);
   OP_DECOMPOSE(cross_entropy_loss);
   OP_DECOMPOSE(arctan2);
+  DECOMPOSE_FUNCTIONAL(diag_embed);
+  DECOMPOSE_FUNCTIONAL(block_diag);
```
Does `vmap(vmap(diag_embed))(x)` work?

bdhirsh replied:
I just tried it locally. It looks like it works:
```
>>> import torch
>>> from functorch import vmap
>>> print(vmap(vmap(torch.diag_embed))(torch.ones(2, 2, 2)))
/scratch/hirsheybar/pytorch-functionalization-env/lib/python3.6/site-packages/functorch-0.0.1a0+20d7754-py3.6-linux-x86_64.egg/functorch/_src/vmap.py:319: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::diagonal_scatter. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at  /raid/hirsheybar/pytorch-functionalization/functorch/functorch/csrc/BatchedFallback.cpp:79.)
  batched_outputs = func(*batched_inputs, **kwargs)
tensor([[[[1., 0.],
          [0., 1.]],

         [[1., 0.],
          [0., 1.]]],


        [[[1., 0.],
          [0., 1.]],

         [[1., 0.],
          [0., 1.]]]])
>>>
```

zou3519 reviewed (commented):
this looks pretty good. We should probably add an explicit test for vmap(vmap(diag_embed)) to check that that works

bdhirsh commented:
I added  `vmap(vmap(torch.diag_embed))` test, and updated the syncing logic to use `sync()` directly now that https://github.com/pytorch/pytorch/pull/68881 landed.
